### PR TITLE
Add replay from block command

### PR DESCRIPTION
**Pull request by PiotrTrzpil**
A CLI command and an API call to replay logs from a given block number.

The result will be that LogBroadcaster will start backfill from the earlier block, but the logs will only be processed only if they were not processed before

### CODE REVIEW

File: core/services/log/broadcaster.go [hunk 1]
```diff
@@ -223,19 +237,19 @@ func (b *broadcaster) startResubscribeLoop() {
 		}
 
 		var backfillFrom null.Int64
-		if b.latestHeadFromDb != nil {
+		if b.requestedBackfillFrom.Valid {
 			// The backfill needs to start at an earlier block than the one last saved in DB, to account for:
```
Comment (connorwstein) [inline core/services/log/broadcaster.go:241]: think this comment needs to be updated?

File: core/services/log/broadcaster.go [hunk 2]
```diff
@@ -306,8 +319,14 @@ func (b *broadcaster) eventLoop(chRawLogs <-chan types.Log, chErr <-chan error)
 			needsResubscribe = b.onRmSubscribers() || needsResubscribe
 
 		case <-debounceResubscribe.C:
-			if needsResubscribe {
-				logger.Debug("LogBroadcaster: returning from the event loop to resubscribe")
+			restartFrom := atomic.LoadInt64(&b.restartFromNumber)
```
Comment (connorwstein) [inline core/services/log/broadcaster.go:322]: maybe even cleaner than piggybacking on the debounceResubscribe and using an atomic int - is we could just use a restart channel and a new case here?

File: core/services/log/broadcaster.go [hunk 3]
```diff
@@ -133,14 +142,19 @@ func (b *broadcaster) Start() error {
 	})
 }
 
-func (b *broadcaster) LatestHead() *models.Head {
-	return b.latestHeadFromDb
+func (b *broadcaster) LatestHeadNumber() null.Int64 {
```
Comment (connorwstein) [inline core/services/log/broadcaster.go:145]: This naming is now a bit confusing but looks like its only used in tests. Lets add a comment

File: core/services/log/broadcaster.go (hunk 2 above)
Comment (connorwstein) [reply core/services/log/broadcaster.go:322]: I think that way we could get rid of restartFromNumber and just use requestedBackfillFrom

Comment (PiotrTrzpil) [reply core/services/log/broadcaster.go:322]: ah yeah it's definitely a more idiomatic go - thanks!

File: core/services/log/broadcaster.go (hunk 1 above)
Comment (PiotrTrzpil) [reply core/services/log/broadcaster.go:248]: actually i left the comment in, but still kept the explicit head from db. It's now a bit simpler i think, and replay is more precise

File: core/services/log/broadcaster.go (hunk 3 above)
Comment (PiotrTrzpil) [reply core/services/log/broadcaster.go]: changed it now to BackfillBlockNumber

File: core/services/log/broadcaster.go [hunk 4]
```diff
@@ -109,6 +116,11 @@ var _ Broadcaster = (*broadcaster)(nil)
 // NewBroadcaster creates a new instance of the broadcaster
 func NewBroadcaster(orm ORM, ethClient eth.Client, config Config, highestSavedHead *models.Head) *broadcaster {
 	chStop := make(chan struct{})
+	//var requestedBackfillFrom null.Int64
```
Comment (connorwstein) [inline core/services/log/broadcaster.go:119]: to remove?

**Review by connorwstein — APPROVED** [review]

**Review by connorwstein — DISMISSED** [review]

**Review by connorwstein — APPROVED** [review]

**Pull request merged**